src/ui/map.js

```diff
@@ -2166,7 +2171,52 @@ class Map extends Camera {
      * @returns {Object} terrain Terrain specification properties of the style.
      */
     getTerrain(): Terrain | null {
-        return this.style.getTerrain();
+        return this.style ? this.style.getTerrain() : null;
+    }
+
+    /**
+     * Sets the fog property of the style.
+     * @param fog The fog properties to set. Must conform the [Mapbox Style Specification](https://docs.mapbox.com/mapbox-gl-js/style-spec/root/#fog).
+     * If `null` or `undefined` is provided, this function call removes the fog from the map.
+     * @returns {Map} `this`
+     * @example
+     * map.setFog({
+     *  "range": [1.0, 12.0],
+     *  "color": 'white',
+     *  "density": 1.0,
+     *  "haze-color": 'rgba(109, 123, 180, 0.5)',
+     *  "sky-blend": 0.1
+     * });
+     */
+    setFog(fog: FogSpecification) {
+        this._lazyInitEmptyStyle();
+        this.style.setFog(fog);
+        return this._update(true);
+    }
+
+    /**
+     * Returns the fog specification or `null` if fog is not set on the map.
+     *
+     * @returns {Object} fog Fog specification properties of the style.
+     */
+    getFog(): Fog | null {
+        return this.style ? this.style.getFog() : null;
+    }
+
+    /**
+     * Returns the fog opacity for a given geolocation.
+     *
+     * An opacity of 0 means that there is no fog contribution for the given geolocation
+     * while a fog opacity of 1.0 means the geolocation is fully obscured by the fog effect.
+     *
+     * If there is no fog set on the map, this function will simply return 0.
+     *
+     * @param {LngLatLike} lnglat The geographical location to evaluate the fog on.
+     * @returns {number} A value between 0 and 1 representing the fog opacity, where 1 means fully within, and 0 means not affected by the fog effect.
+     */
+    getFogOpacity(lnglat: LngLatLike): number {
```

Comment:
The `query-` prefix is used for methods getting map or rendered data. Could `queryFogOpactiy` work for a start?  Would `queryRenderedPixel`  be more flexible, allowing developers to understand things like `Is it sky`, `is it fogy`, `is it raster tile, whats the color`, `whats the elevation`.

Reply:
Updated for `queryFogOpactiy`.